src/server/schema.js

```diff
@@ -42,60 +47,154 @@ const gqlTypeToAggsHistogramName = {
+    ):  [${firstLetterUpperCase(f.key)}]`;
+    Object.keys(f.value.properties).forEach((p) => {
+      if (!existingFields.has(p)) {
+        const field = f.value.properties[p];
+        if (field.type === 'nested') {
+          queueFields.push({ key: p, value: field });
+          existingFields.add(p);
+        }
+      }
+    });
+  }
+  return `${sType}
+  `;
 };
 
+// eslint-disable-next-line max-len
+const getFieldGQLTypeMapForProperties = (esInstance, esIndex, properties) => Object.keys(properties).map((field) => {
+  const esFieldType = properties[field].type;
+  const gqlType = getGQLType(esInstance, esIndex, field, esFieldType);
+  return {
+    field, type: gqlType, esType: esFieldType, props: properties[field].properties,
+  };
+});
+
 const getFieldGQLTypeMapForOneIndex = (esInstance, esIndex) => {
   const fieldESTypeMap = esInstance.getESFieldTypeMappingByIndex(esIndex);
-  const fieldGQLTypeMap = Object.keys(fieldESTypeMap).map((field) => {
-    const esFieldType = fieldESTypeMap[field];
-    const gqlType = getGQLType(esInstance, esIndex, field, esFieldType);
-    return { field, type: gqlType };
+  return getFieldGQLTypeMapForProperties(esInstance, esIndex, fieldESTypeMap);
+};
+
+const getArgsByField = (fieldType, props) => {
+  const keys = Object.keys(props);
+  const baseProps = [];
+  keys.forEach((k) => {
+    if (props[k].type !== 'nested') baseProps.push({ field: k, type: esgqlTypeMapping[props[k].type] });
   });
-  return fieldGQLTypeMap;
+  return `(${baseProps.map(entry => `${entry.field}: [${entry.type}]`).join(',')})`;
+};
+
+const getSchemaType = (fieldType, fieldToArgs) => {
+  if (fieldType.esType === 'nested') {
+    return `${fieldType.field} ${fieldToArgs[fieldType.field]}: ${fieldType.type}`;
+  }
+  return `${fieldType.field}: ${fieldType.type}`;
 };
 
 const getTypeSchemaForOneIndex = (esInstance, esIndex, esType) => {
   const fieldGQLTypeMap = getFieldGQLTypeMapForOneIndex(esInstance, esIndex);
+  const fieldESTypeMap = esInstance.getESFieldTypeMappingByIndex(esIndex);
   const esTypeObjName = firstLetterUpperCase(esType);
-  const typeSchema = `
+  const existingFields = new Set([]);
+  const fieldToArgs = {};
+
+  const queueTypes = [];
+  Object.keys(fieldESTypeMap).forEach((fieldKey) => {
+    const esFieldType = fieldESTypeMap[fieldKey].type;
+    if (esFieldType === 'nested' && !existingFields.has(fieldKey)) {
+      const props = fieldESTypeMap[fieldKey].properties;
+      queueTypes.push({ type: fieldKey, props });
+      existingFields.add(fieldKey);
+      fieldToArgs[fieldKey] = getArgsByField(fieldKey, props);
+    }
+  });
+
+  let sTypeSchema = `
     type ${esTypeObjName} {
-      ${fieldGQLTypeMap.map(entry => `${entry.field}: ${entry.type},`).join('\n')}
+      ${fieldGQLTypeMap.map(entry => `${getSchemaType(entry, fieldToArgs)},`).join('\n')}
     }
   `;
-  return typeSchema;
+
+  while (queueTypes.length > 0) {
+    const t = queueTypes.shift();
```

Comment:
This might be picky but I think we should follow a naming rule that variable name really conveys what it does: `t` here could be renames as `type` probably? or `sType` could named as `singleType`? unless it's a just index (like `i`), let's put more semantic names.